### PR TITLE
Added SpecAugment for the speech recognition task.

### LINKED ISSUE #1835: Implement SpecAugment for speech recognition
## 🚀 Feature Request
Implement [SpecAugment](https://arxiv.org/abs/1904.08779) for the ASR task

### Motivation
SpecAugment is a data augmentation method for ASR that provided SOTA results and is now widely used in the community. It is also used in [this](https://arxiv.org/abs/2002.12231) recent paper by facebook. It would be nice to replicate easily ASR experiments with fairseq.

### Pitch
SpecAugment is a transformation on top of the data. The user should be able to activate it for a speech_recognition training independently from the used model as long as the input is provided as spectrograms.

### Alternatives
There are other data augmentation methods for ASR but specaugment is the most widely used, and so far it proved to be complementary to other methods, not an alternative.

### Additional context
SpecAugment has also been used for direct speech-to-text translation. 
https://arxiv.org/abs/1911.08876
https://zenodo.org/record/3525492

### PR DESCRIPTION
**Pull request by mattiadg**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/pytorch/fairseq/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  

## What does this PR do?
Fixes #1835 .

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: examples/speech_recognition/modules/specaugment.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+def specaugment(mel_spectrogram, frequency_masking_para=27,
+                 time_masking_para=100, frequency_masking_num=1, time_masking_num=1):
+    """Spec augmentation Calculation Function.
+
+    'SpecAugment' have 3 steps for audio data augmentation.
+    first step is time warping using Tensorflow's image_sparse_warp function.
+    Second step is frequency masking, last step is time masking.
+
+    # Arguments:
+      mel_spectrogram(numpy array): audio file path of you want to warping and masking.
```
Comment (mgaido91) [inline examples/speech_recognition/modules/specaugment.py]: it is not a path, but a numpy array containing the filterbanks, isn't it? Actually, my undestanding is that this is a tensor, not a numpy array, am I wrong?

File: examples/speech_recognition/modules/specaugment.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        new_spectrograms = torch.stack(new_spectrograms)
+        batch['net_input']['src_tokens'] = new_spectrograms
+        return batch
+
+
+def specaugment(mel_spectrogram, frequency_masking_para=27,
+                 time_masking_para=100, frequency_masking_num=1, time_masking_num=1):
+    """Spec augmentation Calculation Function.
+
+    'SpecAugment' have 3 steps for audio data augmentation.
```
Comment (mgaido91) [inline examples/speech_recognition/modules/specaugment.py]:
```suggestion
    'SpecAugment' has 3 steps for audio data augmentation.
```

File: examples/speech_recognition/modules/specaugment.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    first step is time warping using Tensorflow's image_sparse_warp function.
```
Comment (mgaido91) [inline examples/speech_recognition/modules/specaugment.py]: are you using tensorflow here?

File: examples/speech_recognition/tasks/speech_recognition.py [hunk 4]
```diff
@@ -77,11 +78,32 @@ def add_args(parser):
         parser.add_argument(
             "--silence-token", default="\u2581", help="token for silence (used by w2l)"
         )
+        parser.add_argument('--specaugment', action='store_true', default=False)
```
Comment (mgaido91) [inline examples/speech_recognition/tasks/speech_recognition.py]: may you please add a `help` here?

File: tests/speech_recognition/test_specaugment.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        self.forward_input = get_dummy_input()
+
+    def test_forward(self):
+        batch = {'net_input': self.forward_input,
+                 'id': torch.arange(self.forward_input['src_tokens'].size(0))
+                 }
+        new_batch = batch.copy()
+        sa = SpecAugment(13, 13, 2, 2, 1.0)
+        new_batch = sa(new_batch)
+        #Verify that not spectrogram values are not changed
```
Comment (mgaido91) [inline tests/speech_recognition/test_specaugment.py]: nit:
```suggestion
        # Verify that not spectrogram values are not changed
```

File: tests/speech_recognition/test_specaugment.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        for k in batch:
+            if k != "net_input":
+                for a, b in zip(batch[k], new_batch[k]):
+                    self.assertEqual(a, b)
+
+        for k in batch['net_input']:
+            #Verify that not spectrogram values are not changed
```
Comment (mgaido91) [inline tests/speech_recognition/test_specaugment.py]: nit:

```suggestion
            # Verify that not spectrogram values are not changed
```

File: examples/speech_recognition/modules/specaugment.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+Related paper : https://arxiv.org/pdf/1904.08779.pdf
+"""
+
+import torch
+import torch.nn as nn
+import numpy as np
+import random
+
+
+class SpecAugment(nn.Module):
```
Comment (mgaido91) [inline examples/speech_recognition/modules/specaugment.py]: may you please add some docs to this class and its methods? Thanks.

Comment (mattiadg) [reply examples/speech_recognition/modules/specaugment.py]: Updated in the last commit

File: examples/speech_recognition/modules/specaugment.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        return batch
+
+
+def specaugment(mel_spectrogram, frequency_masking_para=27,
+                 time_masking_para=100, frequency_masking_num=1, time_masking_num=1):
+    """
+    Spec augmentation Calculation Function.
+    'SpecAugment' has 3 steps for audio data augmentation, but only 2 are implemented in this version.
+    First step is frequency masking, second step is time masking.
+    :param mel_spectrogram: 2-dimensional Tensor representing a spectrogram
```
Comment (mgaido91) [inline examples/speech_recognition/modules/specaugment.py]: in this repo the convention/format used is comments is:
```
Args:
  ....
```
Moreover, may you please also add a description of what each of the 2 dimentions mean? Thanks,

Comment (mattiadg) [reply examples/speech_recognition/modules/specaugment.py]: I hope that it's better now

**Pull request closed without merging**